Python/instrumentation.c

```diff
@@ -2036,6 +2048,36 @@ _PyMonitoring_GetLocalEvents(PyCodeObject *code, int tool_id, _PyMonitoringEvent
     return 0;
 }
 
+int _PyMonitoring_ClearToolId(int tool_id)
+{
+    assert(0 <= tool_id && tool_id < PY_MONITORING_TOOL_IDS);
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+
+    for (int i = 0; i < _PY_MONITORING_EVENTS; i++) {
+        PyObject *func = _PyMonitoring_RegisterCallback(tool_id, i, NULL);
+        if (func != NULL) {
+            Py_DECREF(func);
+        }
+    }
+
+    if (_PyMonitoring_SetEvents(tool_id, 0) < 0) {
+        return -1;
+    }
+
+    _PyEval_StopTheWorld(interp);
+    uint32_t version = global_version(interp) + MONITORING_VERSION_INCREMENT;
+    if (version == 0) {
+        PyErr_Format(PyExc_OverflowError, "events set too many times");
+        _PyEval_StartTheWorld(interp);
+        return -1;
+    }
+    interp->monitoring_tool_versions[tool_id] = version;
```

Comment:
We need to force all code object, executing or not, to "refresh" because the local events will be changed, so we need to update global version for the code objects to know.

Reply:
I think I understood you here. You think it would make sense to do

```c
interp->monitoring_tool_versions[tool_id] = 0;
```

here so all the code objects checking `tool_id` will know they are out of date. However, setting it to the latest global version will have the same effect - none of the code object has this version either.

If we set this to `0`, we need to set it to the latest global version some time in the future, before we use this tool to instrument code again.

So we can either set it to `0` here and set it to `global_version` in `use_tool_id`, or we can do it together here(I can add some. comments here to explain).

Reply:
Ah I was wrong in the previous comment. The update to global version is critical here because the user can set local events immediately after `clear_tool_id`. `clear_tool_id` does not surrender the `tool_id`. So we have to set the tool id version to the latest global version here for the following local events.